src/libstd/ffi/c_str.rs

```diff
@@ -193,20 +193,19 @@ pub struct CString {
 /// [`String`]: ../string/struct.String.html
 /// [`CString`]: struct.CString.html
 /// [`from_ptr`]: #method.from_ptr
-#[derive(Hash)]
 #[stable(feature = "rust1", since = "1.0.0")]
-// FIXME:
-// `fn from` in `impl From<&CStr> for Box<CStr>` current implementation relies
-// on `CStr` being layout-compatible with `[u8]`.
-// When attribute privacy is implemented, `CStr` should be annotated as `#[repr(transparent)]`.
-// Anyway, `CStr` representation and layout are considered implementation detail, are
-// not documented and must not be relied upon.
-pub struct CStr {
-    // FIXME: this should not be represented with a DST slice but rather with
-    //        just a raw `c_char` along with some form of marker to make
-    //        this an unsized type. Essentially `sizeof(&CStr)` should be the
-    //        same as `sizeof(&c_char)` but `CStr` should be an unsized type.
-    inner: [c_char]
+pub struct CStr(CStrExtern);
+
+extern "C" {
+    // HACK: workaround for https://github.com/rust-lang/rust/issues/46665
+    type CStrExtern;
+}
+
+#[stable(feature = "rust1", since = "1.0.0")]
+impl Hash for CStr {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.to_bytes().hash(state)
+    }
```

Comment:
Since it used the derived(Hash) before, it must have included the terminating 0 byte, but in the new Hash it does not. Just to lift the question in code review, that the hash is changing, and if that's ok?

Reply:
Hashes changing is totally fine. They just need to stay equally unique. Adding a `0.hash(state)` to the function body would not create a better hash.

Reply:
We do have a unit test checking that `CStr` and `CString` produce the same hash for the same input. I changed the hashing to also hash the trailing 0 byte.